[38기 김택수] ADD : category 이름 반환 기능 구현

robinkim93 opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x] 기능 추가
- [ ] 리뷰 반영
- [x] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
- category 이름 반환하는 기능 구현 목표
- 데이터 구조변환 방식 리팩토링

<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
1. category 버튼에 hover 될 때, category 이름 받을 수 있도록 기능구현
-  category 및 subcategory 이름 반환하여 화면에 렌더링 할 수 있게 구현

2. 기존 Service단에서 하던 데이터 구조 변환을 Query문으로 해결
-  JSON_OBJECT, JSON_ARRAY, GROUP BY 이용하여 Query문으로 데이터를 변환하여 반환가능하도록 리팩토링
<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)

<br />

## :: 기타 질문 및 특이 사항

robinkim93 committed:
[ADD : category 이름 반환 기능 구현]

lordmyshepherd commented on controller/categoriesController.js:
```diff
@@ -0,0 +1,11 @@
+const categoriesService = require("../service/categoriesService");
+
+const getCategoriesName = async (req, res, next) => {
+  const categories = await categoriesService.getCategoriesName();
+  const subCategories = await categoriesService.getSubCategoriesName();
+  return res
+    .status(200)
+    .json({ categoriesData: categories, subCategoriesData: subCategories });
+};
```
**피드백 1.** 카테고리 이름만 반환하기 보다는 `category` 정보 전체를 반환해서, 추후에 클라이언트에서 카테고리에 해당하는 상품 목록을 호출할 때 `categoryId`를 가지고 호출할 수 있도록 만들어주세요!

lordmyshepherd replied:
**피드백 2.**  데이터 구조도 더욱 명확하게 만들어주는 것이 좋습니다. 이러한 구조로 보내주면 클라이언트에서도 화면을 랜더링할 때 더 쉽게 구현할 수 있겠죠?

```javascript

{
    "data" : [
        {
            "categoryId" : 1, 
            "name" : "A",
            "subCategories" : [
                 {
                      "subCategoryId" : 1, 
                      "name" : "A-a", 
                 },
                 {
                      "subCategoryId" : 2, 
                      "name" : "A-b", 
                 },
            ],
        },
        {
            "categoryId" : 1, 
            "name" : "A",
            "subCategories" : [
                 {
                      "subCategoryId" : 1, 
                      "name" : "A-a", 
                 },
                 {
                      "subCategoryId" : 2, 
                      "name" : "A-b", 
                 },
            ],
        },
        {
            "categoryId" : 1, 
            "name" : "A",
            "subCategories" : [
                 {
                      "subCategoryId" : 1, 
                      "name" : "A-a", 
                 },
                 {
                      "subCategoryId" : 2, 
                      "name" : "A-b", 
                 },
            ],
        },
}


```

lordmyshepherd commented on model/categoriesDao.js:
```diff
@@ -0,0 +1,15 @@
+const myDataSource = require("../util/dataSource");
+
+const getCategoriesName = async () => {
+  return await myDataSource.query(`
+        SELECT * FROM categories
+    `);
+};
+
+const getSubCategoriesName = async () => {
+  return await myDataSource.query(`
+        SELECT * FROM sub_categories
+    `);
+};
```
```suggestion
const getCategories = async () => {
  return await myDataSource.query(`
        SELECT * FROM categories
    `);
};

const getSubCategories = async () => {
  return await myDataSource.query(`
        SELECT * FROM sub_categories
    `);
};
```

robinkim93 committed:
[MODIFY : category 반환 데이터 구조 변경 및 Function naming 변경]

robinkim93 committed:
[MODIFY : conflict 해결]

lordmyshepherd commented on model/categoriesDao.js:
```diff
@@ -0,0 +1,24 @@
+const myDataSource = require("../util/dataSource");
+
+const getCategoriesName = async () => {
+  return await myDataSource.query(`
+        SELECT 
+          c.id AS mainCategoriesId,
+          c.name AS mainCategoriesName
+        FROM categories AS c
+    `);
+};
```
```suggestion
const getCategories = async () => {
  return await myDataSource.query(`
        SELECT 
          c.id AS mainCategoriesId,
          c.name AS mainCategoriesName
        FROM categories AS c
    `);
};
```

lordmyshepherd commented on model/categoriesDao.js:
```diff
@@ -0,0 +1,24 @@
+    `);
+};
+
+const getSubCategoriesName = async (categoriesId) => {
```
```suggestion
const getSubCategories = async (categoriesId) => {
```

lordmyshepherd commented on service/categoriesService.js:
```diff
@@ -0,0 +1,14 @@
+const categoriesDao = require("../model/categoriesDao");
+
+const getCategoriesName = async () => {
+  const mainCategories = await categoriesDao.getCategoriesName();
+
+  for (let i = 0; i < mainCategories.length; i++) {
+    const subCategories = await categoriesDao.getSubCategoriesName(i + 1);
+    mainCategories[i].subCategories = subCategories;
+  }
+
+  return mainCategories;
+};
```
JSON_ARRAY와 JSON_OBJECT를 사용해서 리펙토링 
해보시면 좋을 것 같습니다~

robinkim93 committed:
[MODIFY : category 반환 데이터 구조 리팩토링]
- 기존 service단에서 for문 이용해 데이터 구조를 변환하였으나, 카테고리의 갯수가 많아질수록 비효율적이었음.
- Query문에서 JSON_ARRAY, JSON_OBJECT, GROUP BY, LEFT JOIN 사용해서 데이터 구조 변환하도록 리팩토링
- 변경 후에는 DB단에서 조건에 맞는 데이터만 가져오기 때문에 for문처럼 반복해서 데이터를 불러올 필요가 없어졌음

lordmyshepherd reviewed (approved):
LGTM 👍🏻
